File: BlockUse/BlockUse/HttpTools.m
//
//  HttpTools.m
//  BlockUse
//
//  Created by 古今 on 16/10/10.
//  Copyright © 2016年 夜风易冷. All rights reserved.
//

#import "HttpTools.h"

@interface HttpTools ()
@property (nonatomic, copy) void (^finishedCallBack)(NSString *);
@end
@implementation HttpTools

- (void)loadData:(void (^)(NSString *))finishedCallBack {
    
    self.finishedCallBack = finishedCallBack;
    
    //1.发送网络的异步请求
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        NSLog(@"已经发送了网络请求:%@", [NSThread currentThread]);
        
        //2.回到主线程
        dispatch_sync(dispatch_get_main_queue(), ^{
            NSLog(@"回到主线程:%@", [NSThread currentThread]);
            
            //3.获取数据，并且将数据回调出去
            finishedCallBack(@"json数据");
        });
        
    }) ;
   
}
@end

File: ClosureUse/ClosureUse/HttpTools.swift
//
//  HttpTools.swift
//  ClosureUse
//
//  Created by 古今 on 16/10/10.
//  Copyright © 2016年 夜风易冷. All rights reserved.
//

import UIKit

class HttpTools {
    
    var finishedCallback : ((_ jsonDat : String) -> ())?
    
    //闭包类型:(_ 参数列表) -> (返回值类型)
    //@escaping :逃逸的
    //swift中要求闭包的参数都不能是外部参数(参数前加“_”)
    func loadData(_ finishedCallback: @escaping (_ jsonDat : String) -> ())  {
        
        self.finishedCallback = finishedCallback
        
        //1.发送异步网络请求
        DispatchQueue.global().async {
            print("发送异步网络请求:\(Thread.current)")
            
            //2.回到主线程
            DispatchQueue.main.sync {
                print("回到主线程:\(Thread.current)")
                
                //3.进行回调
                finishedCallback("json数据")
            }
        }
        
    }
}
